#  Pseudobulk and related approaches for scRNA-seq analysis

## Learning Objectives

- Understanding considerations for when to use different DGE algorithms on scRNA-seq data
- Using FindMarkers to evaluate significantly DE genes
- Aggregating single cell expression data into a pseudobulk counts matrix to run a DESeq2 workflow
- Evaluating expression patterns of differentially expressed genes at the pseudobulk and single cell level
- Application of methods for evaluating differential proportions of cells between conditions

## Installations

### On your desktop

1. [R](https://www.r-project.org/)
2. [RStudio](https://posit.co/download/rstudio-desktop/)
4. [The listed R packages](../README.md#installation-requirements)

## Lessons

1. [Introduction to scRNA-seq](https://hbctraining.github.io/scRNA-seq_online/lessons/01_intro_to_scRNA-seq.html)
2. [scRNA-seq: From sequence reads to count matrix](https://hbctraining.github.io/scRNA-seq_online/lessons/02_SC_generation_of_count_matrix.html)
3. [scRNA-seq: From counts to clusters](../lessons/00_counts_to_clusters_overview.md)
4. [Project setup and data exploration](../lessons/01_setup_intro_dataset.md)
5. [Differential expression analysis using `FindMarkers()`](../lessons/02_DEanalysis_using_FindMarkers.md)
6. [Aggregating counts by celltype using pseudobulk approach](../lessons/03_pseudobulk_DESeq2.md)
7. [DE analysis of pseudobulk data using DESeq2](../lessons/04_pseudobulk_DE_analysis.md)
8. [Visualization of differentially expressed genes](../lessons/05_pseudobulk_DE_visualizations.md)
9. [Comparison of results from different DE approaches](../lessons/06_DE_comparisons.md)
10. [Functional Analysis](../lessons/07_functional_analysis_pseudobulk.md)
11. [Methods for Differental Abundance](../lessons/08_differential_abundance.md)


## Answer key

- [Day 1 exercises](../homework/Day_1_answer_key.md)
- [Day 2 exercises](../homework/Day_2_answer_key.md)
***

*These materials have been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*
